fix ansible+vagrant+virtualbox+centos7 configuration

rutsky opened the pull request:
Changes in Ansible + Vagrant configuration:

* By default on VirtualBox public interface is not `eth0`, but `eth1`: I automatically use proper interface for flannel and when generating `/etc/hosts`.

  This should fix #586 .

* Fix `bin_dir` on CentOS with K8S installed from packages.

These changes are *almost* enough to start K8S cluster in Vagrant on VirtualBox + CentOS 7.
Unfortunately CentOS 7 yet have only [1.2.0alpha1](https://git.centos.org/summary/rpms!kubernetes.git) which doesn't include [commit with `successThreshold` and `failureThreshold` liveness probe parameter](https://github.com/kubernetes/kubernetes/commit/1e88a682da871a4a7f811dc5efcfe3890be8bbf8). This leads to errors in `kube-addons` like following one:

```
мар 24 09:43:38 kube-master kube-addons.sh[30042]: error validating "/etc/kubernetes/addons/dns/skydns-rc.yaml": error validating data: [found invalid field successThreshold for v1.Probe, found invalid field failureThreshold for v1.Probe]; if you choose to ignore these errors, turn validation off with --validate=false
```

To bypass this issue I locally patched kube-addons scripts to include `--validate=false` in `kubectl` invocations:

```
diff --git a/ansible/roles/kubernetes-addons/files/kube-addon-update.sh b/ansible/roles/kubernetes-addons/files/kube-addon-update.sh
index ede2832..ab0e023 100755
--- a/ansible/roles/kubernetes-addons/files/kube-addon-update.sh
+++ b/ansible/roles/kubernetes-addons/files/kube-addon-update.sh
@@ -259,7 +259,7 @@ function create-object() {
     log INFO "Creating new ${obj_type} from file ${file_path} in namespace ${namespace}, name: ${obj_name}"
     # this will keep on failing if the ${file_path} disappeared in the meantime.
     # Do not use too many retries.
-    run-until-success "${KUBECTL} create --namespace=${namespace} -f ${file_path}" ${NUM_TRIES} ${DELAY_AFTER_ERROR_SEC}
+    run-until-success "${KUBECTL} create --validate=false --namespace=${namespace} -f ${file_path}" ${NUM_TRIES} ${DELAY_AFTER_ERROR_SEC}
 }
 
 function update-object() {
diff --git a/ansible/roles/kubernetes-addons/files/kube-addons.sh b/ansible/roles/kubernetes-addons/files/kube-addons.sh
index db555fe..916f4e4 100644
--- a/ansible/roles/kubernetes-addons/files/kube-addons.sh
+++ b/ansible/roles/kubernetes-addons/files/kube-addons.sh
@@ -118,7 +118,7 @@ function create-resource-from-string() {
   local -r config_name=$4;
   local -r namespace=$5;
   while [ ${tries} -gt 0 ]; do
-    echo "${config_string}" | ${KUBECTL} --namespace="${namespace}" create -f - && \
+    echo "${config_string}" | ${KUBECTL} --validate=false --namespace="${namespace}" create -f - && \
         echo "== Successfully started ${config_name} in namespace ${namespace} at $(date -Is)" && \
         return 0;
     let tries=tries-1;
```

Note that I tested with `source_type: packageManager` in `group_vars/all.yml`.

Cluster starts with:

```
$ vagrant up --provider virtualbox
[...]
PLAY RECAP *********************************************************************
kube-master                : ok=260  changed=78   unreachable=0    failed=0   
kube-node-1                : ok=125  changed=39   unreachable=0    failed=0   
kube-node-2                : ok=124  changed=39   unreachable=0    failed=0 
```

On kube-master:

```
[root@kube-master ~]# kubectl get nodes
NAME          LABELS                               STATUS    AGE
kube-node-1   kubernetes.io/hostname=kube-node-1   Ready     8m
kube-node-2   kubernetes.io/hostname=kube-node-2   Ready     8m
[root@kube-master ~]# kubectl get pods --all-namespaces
NAMESPACE     NAME                                   READY     STATUS             RESTARTS   AGE
kube-system   elasticsearch-logging-v1-dq4kh         1/1       Running            0          9m
kube-system   elasticsearch-logging-v1-hkymz         1/1       Running            0          9m
kube-system   fluentd-elasticsearch-kube-node-1      1/1       Running            0          7m
kube-system   fluentd-elasticsearch-kube-node-2      1/1       Running            0          6m
kube-system   kibana-logging-v1-irn4b                0/1       CrashLoopBackOff   5          9m
kube-system   kube-dns-v11-0u775                     3/4       Running            3          9m
kube-system   monitoring-influxdb-grafana-v3-clyo8   2/2       Running            0          9m
```

I haven't tested that configured cluster is actually working (see failing kibana-logging above), but at least it looks like properly configured.

My goal is to setup cluster on CoreOS in Vagrant and these changes are required for this anyway (e.g. proper network interface configuration).

danehans commented on ansible/roles/common/tasks/main.yml:
```diff
@@ -31,6 +31,12 @@
+  set_fact:
+    bin_dir: "/usr/bin"
+  when: ansible_distribution == "CentOS" and source_type == "packageManager"
+
```
@rutsky this might not be the best place to fix bin_dir. @eparis can you confirm if you want the project to follow [1] for the default bin_dir?

[1] https://github.com/kubernetes/contrib/tree/master/init

rutsky replied:
@danehans @eparis I would love to see `/usr/bin` as default `bin_dir`.

I used other approach here because I'm not sure how to properly change `bin_dir` without breaking other configurations which I can't test.

I will rebase this PR on fix of `bin_dir`, if there will be any (I can try to fix it myself, if nobody volunteers).

danehans replied:
@rutsky let's get confirmation from @eparis first. Thanks.

danehans commented:
Link to similar PR to address the default bin_dir issue: https://github.com/kubernetes/contrib/pull/618#issuecomment-200903100

atosatto commented on ansible/vagrant/vagrant-ansible.yml:
```diff
@@ -1,19 +1,45 @@
         dest=/etc/hosts
         regexp=".*{{ item }}$"
-        line="{{ hostvars[item].ansible_default_ipv4.address }} {{item}}"
+        line="{{ hosts_public_ips[item] }} {{item}}"
```
Isn't that too much complex? 
I think we can achieve this using a custom fact like `public_iface`that is set by default to `ansible_default_ipv4`and to `ansible_eth1` in case the provider is Virtualbox and keep using the "Build hosts file" and "Remove hostname from localhost line" tasks with straightforward changes

```
- name: "Build hosts file"
      lineinfile:
        dest=/etc/hosts
        regexp=".*{{ item }}$"
        line="{{ hostvars[item][public_iface]['ipv4']['address']}} {{item}}"
        state=present
      when: hostvars[item][public_iface]['ipv4']['address'] is defined
      with_items: groups['all']
    - name: "Remove hostname from localhost line"
      replace:
        dest=/etc/hosts
        regexp="{{ item }} localhost?"
        replace="localhost"
      when: hostvars[item][public_iface]['ipv4']['address'] is defined
      with_items: groups['all']
```

atosatto replied:
For sake of completeness we can rewrite the playbook in the following way

```yml
- hosts: all
  sudo: yes
  defaults:
    - public_iface: ansible_eth1
  tasks:
    - name: "Build hosts file"
      lineinfile:
        dest=/etc/hosts
        regexp=".*{{ item }}$"
        line="{{ hostvars[item][public_iface]['ipv4']['address']}} {{item}}"
        state=present
      when: hostvars[item][public_iface]['ipv4']['address'] is defined
      with_items: groups['all']
    - name: "Remove hostname from localhost line"
      replace:
        dest=/etc/hosts
        regexp="{{ item }} localhost?"
        replace="localhost"
      when: hostvars[item][public_iface]['ipv4']['address'] is defined
      with_items: groups['all']
````
and ovveride the `public_iface` using the Vagrant's `extra_vars` only in the case of Virtualbox.

@eparis and @ingvagabund what do you think about it?

PS: I've not syntactically checked the playbook snippet but it should works!

atosatto replied:
BWT, your solution allows us to keep using the `ansible_default_ipv4` fact while my proposal requires to explicitly refer to an interface using the `ansible_eth0`, etc facts... :)

rutsky replied:
@atosatto I like your approach and I would prefer to do it in your way.

However I used another approach because your approach is not backward compatible: `ansible_default_ipv4` is *default* interface for routing packets, not definitely `eth0`.
Also `eth0` is not even "first" network interface, interfaces may have arbitrary numbers (e.g start from `eth1`) or be like `enp2s0`.

@atosatto, @eparis, @ingvagabund If this is not significant at this moment (because most probably default interface will be `eth0` on used images) I would prefer to adapt @atosatto approach.

atosatto replied:
Whoah @rutsky! :) I'm so happy you liked my suggestion.

I believe that the semantics we should to these ifaces is something involving clusterization and orchestrations. In that sense I dont't feel the `ansible_default_ipv4` the best fact to use in that sense.. ;)

eparis commented:
I just merged the default `bin_dir` change. We'll see who that breaks. Let me know when you think this is ready for merge.

danehans commented:
@rutsky can you update this PR to remove 430f81af71317fc211f0efd8eca6e75e939441b5 as this fix was implemented in https://github.com/kubernetes/contrib/commit/a5608e64cc14ce0b8a82ba5785be095bc7bb2ef0

rutsky commented:
@danehans
> @rutsky can you update this PR to remove 430f81a as this fix was implemented in a5608e6

Will do.

@eparis
> I just merged the default bin_dir change. We'll see who that breaks. Let me know when you think this is ready for merge.

Can you take a look at @atosatto suggestion in this PR about introducing `public_iface` and use it instead of `ansible_default_ipv4`? I would like to fix this too, just want your approval of the idea.

rutsky commented:
@danehans, @eparis, @atosatto, I rebased and reworked `vagrant-ansible.yml` according to @atosatto suggestion, please take a look.

Now `vagrant-ansible.yml` uses `public_iface` variable to select public Ethernet interface (`eth0` by default and on VirtualBox set to `eth1`).

danehans commented:
@rutsky i'm giving the vagrant stuff a spin for the first time. I see the following error in my vagrant unusable run:
```
TASK [flannel : Start flannel] *************************************************
changed: [kube-master]
fatal: [kube-node-2]: FAILED! => {"changed": false, "failed": true, "msg": "Job for flanneld.service failed because a timeout was exceeded. See \"systemctl status flanneld.service\" and \"journalctl -xe\" for details.\n"}
fatal: [kube-node-1]: FAILED! => {"changed": false, "failed": true, "msg": "Job for flanneld.service failed because a timeout was exceeded. See \"systemctl status flanneld.service\" and \"journalctl -xe\" for details.\n"}
```

Let me try your patch and see if it fixes the issue.

rutsky commented:
@danehans please also take a look at my other edits described in first message of this PR: I used  `source_type: packageManager`, and kunernetes on stock centos7 is not compatible with new DNS add-on manifest (this can be workarounded as I described above).

danehans commented:
@rutsky Looks like I get past the Flannel errors after applying the commits in this PR. Now my vagrant setup fails here:
```
TASK [master : Copy master binaries] *******************************************
failed: [kube-master] => (item=kube-apiserver) => {"failed": true, "item": "kube-apiserver", "msg": "could not find src=/Users/daneyonhansen/Desktop/_output/local/go/bin/kube-apiserver"}
failed: [kube-master] => (item=kube-scheduler) => {"failed": true, "item": "kube-scheduler", "msg": "could not find src=/Users/daneyonhansen/Desktop/_output/local/go/bin/kube-scheduler"}
failed: [kube-master] => (item=kube-controller-manager) => {"failed": true, "item": "kube-controller-manager", "msg": "could not find src=/Users/daneyonhansen/Desktop/_output/local/go/bin/kube-controller-manager"}
failed: [kube-master] => (item=kubectl) => {"failed": true, "item": "kubectl", "msg": "could not find src=/Users/daneyonhansen/Desktop/_output/local/go/bin/kubectl"}
failed: [kube-master] => (item=kubelet) => {"failed": true, "item": "kubelet", "msg": "could not find src=/Users/daneyonhansen/Desktop/_output/local/go/bin/kubelet"}
```
Again, this is my first time working with the vagrant environment. Am I supposed to download the kube bins to my laptop prior to doing vagrant up? I am running vagrant/vbox on Mac (10.10.5).
```
$ ansible --version
ansible 2.0.1.0
  config file = /Users/daneyonhansen/Desktop/contrib/ansible/vagrant/ansible.cfg
  configured module search path = Default w/o overrides
```

danehans commented:
@rutsky I just saw your message about source_type. Let me make the edits and try again. Thanks!

eparis commented on ansible/vagrant/vagrant-ansible.yml:
```diff
@@ -1,19 +1,20 @@
 - hosts: all
   sudo: yes
+  vars:
+    - public_iface: eth0
```
Sorry! I should have said something. No way. Most modern computers don't even have an eth0.  Does ansible have any facts we could use? We've got to find something other than eth0.

eparis replied:
Maybe use `ansible_default_ipv4.interface`?

atosatto replied:
@eparis great suggestion! :) In this way we can also be consistent with the legacy default. Nice!

eparis replied:
That fact won't be defined here though, so it'll be a little more of a pain to get it right...

rutsky replied:
@eparis Are you sure? I tested that by default `eth0` is used in my environment, but I'm not sure is Ansible cached some facts from previous runs (I don't explicitly use any facts caching, but not sure is Ansible does any).

In this part Ansible *will* gather facts, but I'm unsure whether this will be done before evaluating `vars:` part or after:

```
- hosts: all
  sudo: yes
  vars:
    - public_iface: "{{ ansible_default_ipv4.interface }}"
  tasks:
    - ...
```

If my tests are correct, it gather facts before evaluating `vars:` (at least in my version of Ansible), and these changes works correctly.

rutsky commented:
@atosatto, @eparis I use `ansible_default_ipv4.interface` as default interface now, thanks!

eparis commented:
If it works, I'm happy...

eparis commented:
@rutsky if anything goes wrong, do you agree to help figure out how to fix it?

rutsky commented:
@eparis sure!

However I don't think CentOS 7 image works very well with their currently bundled old K8S alpha version.
I want to use Core OS with K8S 1.2.0 in Vagrant and these fixes will be needed for Core OS too (since they are related to Vagrant and Virtualbox, not to specific OS release), and I will support them.

BTW, if there are options to write actual e2e tests for this Ansible part of contrib repo, I would like to add configuration that I will use (Ansible + Vagrant) to prevent it from breaking in the future.

eparis commented:
I don't think we have a reasonable way to e2e these, especially this part. @ingvagabund will get mad at us when we break things, since he uses the ansible stuff in the Red Hat CI system. Maybe we'll get there...

danehans commented:
@rutsky fyi... I was able to deploy my vagrant env using your patch using centos7 and ```source_type: packageManager```

cc @stephenrlouie

stephenrlouie commented:
@rutsky I'm hoping to help in the effort to get CoreOS vagrant support upstream. 

I'm working with @danehans and have been building out a virtual dev environment for CoreOS. I have leveraged [coreos-vagrant] (https://github.com/stephenrlouie/coreos-vagrant) and a ["build_master"] (https://github.com/stephenrlouie/PXE-Boot-VM/tree/gemini) by cherry-picking some work from @danehans. 

I have been able to create a working K8s environment on a virtualbox provider, just need to plug into the /contrib/ansible/vagrant work and get it upstream. 

@rutsky: feel free to reach out to me on slack (kubernetes-users) @ stelouie for extra support. Now that I have my own virtual env up, I'm going to try and integrate into the /contrib/ansible/vagrant work

danehans commented:
I created https://github.com/kubernetes/contrib/issues/636 b/c source_type: localBuild still does not work in Vagrant. I expect this to be an issue with cores since bins are pulled from GH.

cc @stephenrlouie

rutsky commented:
@stephenrlouie, @danehans, I would love to unite our efforts on upstream support for Vagrant+Ansible+CoreOS configuration.
Please take a look at #644 with initial support of CoreOS image in Vagrant+Ansible. There are still some issues (at least with Flannel) so if you will fix them before I do I would love to review them.

danehans commented:
@rutsky that's exactly what I was thinking. You can find us on the #kubernetes-users slack channel: @daneyon @stepehenrlouie @adamscha. Ping us on slack when you have a moment so we know your nic.

@stephenrlouie is leading our dev/cicd environment. We are still in the early stages of organizing our project. @stephenrlouie will be comparing our existing coreos/kube vagrant setup [1][2] with contrib/ansible/vagrant. IMO we will transition our work in [2] to contrib/ansible/vagrant and keep [1] separate.

[1] https://github.com/stephenrlouie/PXE-Boot-VM/commits/gemini
[2] https://github.com/stephenrlouie/coreos-vagrant/tree/gemini

Looking forward to working more closely with one another.

rutsky commented:
@danehans I registered on #kubernetes-users in Slack with "rutsky" nickname, feel free to ping me there (though I don't expect to be very active there).
